cmd/system-shutdown: Rename die() -> die_reboot() to fix build failures with LTO enabled

dcermak opened the pull request:
`void die(const char *msg)` from `cmd/system-shutdown/system-shutdown-utils.h` clashes with `void die(const char *fmt, ...)` from `cmd/libsnap-confine-private/utils.h`, which turns into an error when you build snap with link time optimization enabled.

This patch renames `die` to `die_reboot`, to make it clearer what it does and to resolve this ambiguity.

chipaca commented:
why're we linking the lib and not using the functions from it? IIRC shutdown-utils was a subset of the ones from snap-confine, split out for expediency? or am i misremembering. My point being, if things have changed and we're linking the lib _anyway_, why aren't we using it? What's -utils even doing?

zyga commented:
@chipaca I'm getting to the bottom of this.

dcermak commented:
> I'd love to know more about what kind of error you saw, and what kind of liner optimisation tricks were utilised.

The error is this one:
```
gcc -Wall -Wextra -Wmissing-prototypes -Wstrict-prototypes -Wno-missing-field-initializers -Wno-unused-parameter -Werror -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -flto=12 -static -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -flto=12  -static  -o system-shutdown/system-shutdown system-shutdown/system_shutdown-system-shutdown-utils.o system-shutdown/system_shutdown-system-shutdown.o libsnap-confine-private.a
libsnap-confine-private/utils.h:25:6: error: type of ‘die’ does not match original declaration [-Werror=lto-type-mismatch]
   25 | void die(const char *fmt, ...);
      |      ^
system-shutdown/system-shutdown-utils.c:57:6: note: types have different parameter counts
   57 | void die(const char *msg)
      |      ^
system-shutdown/system-shutdown-utils.c:57:6: note: ‘die’ was previously declared here
lto1: all warnings being treated as errors
lto-wrapper: fatal error: /usr/bin/gcc returned 1 exit status
compilation terminated.
/usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:1724: system-shutdown/system-shutdown] Error 1
```
There's really no tricks involved at all, beside using openSUSE's standard build flags:
```
export CFLAGS="-O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -flto=$(nproc)"
```

> 
> I had a look at how we link `system-shutdown` and there's one thing that worries me, we link with `libsnap-confine-private.a`. This also defines `die` so it's clear (deliberately or not) we have two definitions of `die` here. With this patch they will be separate and `system-shutdown` can now die with reboot or without one.
> 
> Hmmm, I wonder how we missed this before.

You won't get a build failure by default due to this unless you have unity builds setup.

dcermak commented:
Please don't merge this yet as I forgot to sign the CLA...

(Or recreate this patch from scratch and close this PR)

zyga commented:
@dcermak I'm looking at fixing this so that die() can be used from both places and to be appropriate in both places, therefore not showing the error you are seeing.

I'm using Tumbleweed as my daily driver and since I obviously have not seen this error before I wonder if we should look at our hardening flags again. Thank you for raising this!

dcermak commented:
> I'm using Tumbleweed as my daily driver and since I obviously have not seen this error before I wonder if we should look at our hardening flags again. Thank you for raising this!

You won't see this unless you are building in a rpmbuild environment (e.g. on OBS, where system:snappy/snapd fails to build with the above error).

zyga commented:
Hopefully #7362 is convincing enough that it can be merged and this issue can be resolved. Thank you for the report and I'll hope to fix this for the next release.

dcermak commented:
I've rebased my branch and fixed the commit message, but wasn't able to sign the CLA yet as I need approval from legal (the responsible person is currently on vacation for another week).

bboozzoo commented:
Renamed the title to make our PR title checker happy again.

bboozzoo commented:
@mvo5 once CLA is sorted out, perhaps we could land this one and then #7362 on top?

zyga commented:
I think 7362 is the proper way out and should land soon given that mvo has changed his mind. I'm inclined to close this PR now.

zyga commented:
I'm closing this since the duplication has been removed.
